File: DirectProgramming/C++/Jupyter/OpenMP-offload-training/datatransfer/target_data_region_solution.cpp
// Solution Using Target Data
#pragma omp target data map(tofrom : x)
{
#pragma omp target map(to : y)
  {
    for (int i = 0; i < ARRAY_SIZE; i++) x[i] += y[i];
  }

  init2(y, ARRAY_SIZE);

#pragma omp target map(to : y)
  {
    for (int i = 0; i < ARRAY_SIZE; i++) x[i] += y[i];
  }
}


// Solution Using Target Enter/Exit/Update
#pragma omp target enter data map(to : x) map(to : y)
#pragma omp target
{
  for (int i = 0; i < ARRAY_SIZE; i++) x[i] += y[i];
}

init2(y, ARRAY_SIZE);

#pragma omp target update to(y)

#pragma omp target
{
  for (int i = 0; i < ARRAY_SIZE; i++) x[i] += y[i];
}
#pragma omp target exit data map(from : x)

File: DirectProgramming/Fortran/Jupyter/OpenMP-offload-training/datatransfer/target_data_region_solution.f90
! Solution Using target data
!$omp target data map(tofrom: x)
    !$omp target map(to:y)
    do i=1, ARRAY_SIZE
        x(i) = x(i) + y(i)
    end do
    !$omp end target

    call init (y, ARRAY_SIZE, 2.0)

    !$omp target map(to:y)
    do i=1, ARRAY_SIZE
        x(i) = x(i) + y(i)
    end do
    !$omp end target
!$omp end target data




! Solution Using target enter/exit/update data
!$omp target enter data map(to:x) map(to:y)
!$omp target
do i=1, ARRAY_SIZE
        x(i) = x(i) + y(i)
end do
!$omp end target

call init (y, ARRAY_SIZE, 2.0)
!$omp target update to(y)

!$omp target
do i=1, ARRAY_SIZE
        x(i) = x(i) + y(i)
end do
!$omp end target
!$omp target exit data map(from:x)
